Unittester failures summary

ZuleykhaPavlichenkova-TomTom opened the pull request:
Currently when we have a failure in the CI (running unit tests), the actual test failure is hidden in mountains of successful runs.

We should print the output of all failing tests again at the end of running the unittester so we can more easily find the failing tests. This can make it easier to look through the CI.

This PR adds the Failures Summary printed out right in the end of the logs by default.
Passing --summarize-failures flag to the `./build/release/test/unittest` should enable summary output in the end of test run. There is a python script to run unit tests one by one (`scripts/run_tests_one_by_one.py`) which should print summary by default.

Python script captures the whole failure details message easily, but the `unittest` - only the failing case with the problematic line number.

Should look like this for `unittest`:
```
=====================================================
================  FAILURES  SUMMARY  ================
=====================================================

1: test/sql/aggregate/aggregates/histogram_table_function.test: 20 
2: test/sql/copy/csv/test_date.test: 11 
```

And for `scripts/run_tests_one_by_one.py`:
```
=====================================================
================  FAILURES  SUMMARY  ================
=====================================================

1: ['test/sql/aggregate/aggregates/histogram_table_function.test']
FAILED WITH ERROR:
 ================================================================================
Wrong result in query! (test/sql/aggregate/aggregates/histogram_table_function.test:20)!
================================================================================
SELECT * FROM histogram_values(integers, i, bin_count := 2);
================================================================================
Mismatch on row 1, column count(index 2)
1 <> 1q
================================================================================
Expected result:
================================================================================
60      1q
80      0
100     1
================================================================================
Actual result:
================================================================================
60      1
80      0
100     1

2: ['test/sql/copy/csv/test_date.test']
FAILED WITH ERROR:
 ================================================================================
Wrong result in query! (test/sql/copy/csv/test_date.test:11)!
================================================================================
COPY date_test FROM 'data/csv/test/date.csv';
================================================================================
Mismatch on row 1, column Count(index 1)
1 <> 1q
================================================================================
Expected result:
================================================================================
1q
================================================================================
Actual result:
================================================================================
1
```

ZuleykhaPavlichenkova-TomTom committed:
Add new --summarize-failures argument to the unittester. It will allow the unittester to print the failed test names + failed line as a run summary

ZuleykhaPavlichenkova-TomTom committed:
add summary to py script

ZuleykhaPavlichenkova-TomTom committed:
outputs one line

ZuleykhaPavlichenkova-TomTom committed:
after merge

ZuleykhaPavlichenkova-TomTom committed:
print failures summary

ZuleykhaPavlichenkova-TomTom committed:
format-fix

ZuleykhaPavlichenkova-TomTom committed:
format-fix

ZuleykhaPavlichenkova-TomTom committed:
summary format changes

ZuleykhaPavlichenkova-TomTom committed:
output test name without brackets

Mytherin commented:
Thanks for the PR! LGTM - the CI failure is unrelated.

Can we start enabling these flags on the various CI runs as well?

Tmonster reviewed (commented):
Looks good! Just the one question

ZuleykhaPavlichenkova-TomTom committed:
collect the error output into the log_message variable and then save to the test file
changed returning types for some of the logger methods

ZuleykhaPavlichenkova-TomTom committed:
add "next case" to the failures summary file so we could have an order number before each test case name in the summary

hmeriann commented:
Made the summary look the same as summary of Python runner results - for now it's only for CompareValues(). Will add more

ZuleykhaPavlichenkova-TomTom committed:
move summary collection method to the SQLLogicTestLogger, add failures collection from UnexpectedStatement

ZuleykhaPavlichenkova-TomTom committed:
collect logs from UnexpectedFailure, PrintResultError

ZuleykhaPavlichenkova-TomTom committed:
add summary logger to CheckQueryResult

ZuleykhaPavlichenkova-TomTom committed:
add summary logger to CheckStatementResult

ZuleykhaPavlichenkova-TomTom committed:
add summary logger to MatchesRegex

ZuleykhaPavlichenkova-TomTom committed:
format fix

ZuleykhaPavlichenkova-TomTom committed:
fixed an issue with extra new line in expected result, cleaned up

hmeriann commented:
there is the `SQLLogicTestLogger` which is used to output all the `unittest` test results. I've added a `SQLLogicTestLogger::AddToSummary(string log_message)` method to write found failure into a file.

I've also modified the return types of most of the `SQLLogicTestLogger` methods so that the strings they return can be collected into a `log_message` variable.

I think that the filename should be defined somewhere else, not in the `SQLLogicTestLogger::AddToSummary(string log_message)` method, maybe globally, like the `duckdb_unittest_tempdir` name.

ZuleykhaPavlichenkova-TomTom committed:
add missing import

hmeriann commented:
We still will have to scroll in case of long logs [like this](https://github.com/duckdb/duckdb/actions/runs/14105375963/job/39510650010?pr=16833#step:9:620)

Tmonster commented:
> We still will have to scroll in case of long logs [like this](https://github.com/duckdb/duckdb/actions/runs/14105375963/job/39510650010?pr=16833#step:9:620)

I was going to say yea. I think if you summarize everything, it's a bit too long of a message. The reason we want this is just to see what files have failing test cases. It's much nicer to look at the files, then run the tests locally to find out exactly what has failed. If you want to have a summary with also the failures, then I suggest having two flags `--summarize-failures` and `--summarize-failues-verbose`, with verbose also publishing expected error messages etc.

Mytherin commented on .github/workflows/Main.yml:
```diff
@@ -193,7 +193,7 @@ jobs:
     - name: Test
       shell: bash
-      run: build/reldebug/test/unittest --force-reload --force-storage
+      run: build/reldebug/test/unittest --force-reload --force-storage --summarize-failures
```
Instead of making this a command line option maybe we can make it an environment variable? That way this also works with the CI runs that run e.g. `make allunit`.

ZuleykhaPavlichenkova-TomTom committed:
set SUMMARIZE_FAILURES env variable

ZuleykhaPavlichenkova-TomTom committed:
get SUMMARIZE_FAILURES value from env
is still possible to run unittest passing --summarize-failures flag to get FAILURES SUMMARY, if an environment variable SUMMARIZE_FAILURES not set

ZuleykhaPavlichenkova-TomTom committed:
add a newline before the next failure in the summary

ZuleykhaPavlichenkova-TomTom committed:
cleaning up git history

carlopi commented on CMakeLists.txt:
```diff
@@ -536,6 +536,9 @@ if(FORCE_ASSERT)
   set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -DDUCKDB_FORCE_ASSERT")
 endif()
 
+# sets SUMMARIZE_FAILURES env variable for unittest
+set(SUMMARIZE_FAILURES_ENV "$ENV{SUMMARIZE_FAILURES}")
+
```
It's this still used?

hmeriann replied:
yes, I've added it to get the `SUMMARIZE_FAILURE` flag from CI env

carlopi replied:
I guess a rephrase is: is `SUMMARIZE_FAILURES_ENV` ever used? I see it only once in the code to set the variable, so I am not sure who reads this

hmeriann replied:
ah, I see! it should be used by `unittest` https://github.com/duckdb/duckdb/blob/efbb8ef88dc0550cd541c84bf2a4fa8410fe7dbb/test/unittest.cpp#L40

carlopi replied:
I think just removing `SUMMARIZE_FAILURES_ENV` entirely is good, and then at runtime either env variable `SUMMARIZE_FAILURES` OR options decide the behaviour

carlopi commented on scripts/run_tests_one_by_one.py:
```diff
@@ -61,6 +66,7 @@ def valid_timeout(value):
 assertions = args.no_assertions
 time_execution = args.time_execution
 timeout = args.timeout
+summarize_failures = args.summarize_failures
```
Should this be ON also if ENV variable?

Either to be handled here or in the summary later (otherwise if ENV is set, python will skip printing)

hmeriann replied:
You're right! Thank you for pointing on this. I think the solution is to pass the ENV variable from the Python script

carlopi commented on test/unittest.cpp:
```diff
@@ -24,18 +26,32 @@ bool TestMemoryLeaks() {
 
 int main(int argc, char *argv[]) {
 	duckdb::unique_ptr<FileSystem> fs = FileSystem::CreateLocal();
 	string test_directory = DUCKDB_ROOT_DIRECTORY;
+	const char *filename("failures_summary.txt");
```
Where the failures should be saved to?
Should that be configurable?
Currently if you do
```
./build/release/test/unittester
```
I guess this will be in your main duckdb folder.

And this means running unittester multiple times will (if I understand this correcty) not delete previous results. I guess both sequential but also parallel are somewhat relevant.

I would think adding UUID can be handy and also allowing to change base folder maybe? or use the folder used by unittester already? or use a variation (|| "_error_summaries") on  the folder used to store data?

Mytherin replied:
I don't really understand why we are writing to a file now in the first place? It seems this file is only used internally (we write to it and then we read it again in the end). Can we keep this state in-memory instead?

ZuleykhaPavlichenkova-TomTom committed:
Since the python runner should always print Failures Summary, it can be set as an ENV variable and passed to the unittest

Mytherin commented on test/sqlite/sqllogic_test_logger.cpp:
```diff
@@ -182,108 +202,138 @@ void SQLLogicTestLogger::OutputHash(const string &hash_value) {
+	log_message += PrintHeader("Actual result:");
+	log_message += PrintLineSep();
 	result.Print();
+	log_message += result.ToString();
```
It feels like we are doing a lot of duplicate work here which complicates the code (`result.Print()` followed by `result.ToString()`).

Generally the way we would achieve either printing or collecting a string is to write to a stream object, e.g. an `std::ostream &ss`. Then we can pass in a `std::stringstream` to obtain a string, or pass in another output stream (e.g. `stdout`) to render to a different location.

Can we rework this to use that construct to avoid having to change/duplicate all of this code?

hmeriann replied:
I thought I could only do it if I touch the `Printer::Print`, but after discussion with @Tmonster I understood how can work around

ZuleykhaPavlichenkova-TomTom committed:
remove --simmarize_failures argument, because we'd like to have FAILURES SUMMARY every time im Python

ZuleykhaPavlichenkova-TomTom committed:
add GetSummary() global to accumulate error logs to variable of type stringstream.
add counter for the failing tests cases to use in the output.

ZuleykhaPavlichenkova-TomTom committed:
revert changes to duckdb::Printer

ZuleykhaPavlichenkova-TomTom committed:
remove unused function.
instead of writing failures summary logs into a file write them to the stringstream object
use GetSummary() function to write failures logs and to output the summary

ZuleykhaPavlichenkova-TomTom committed:
clean up

ZuleykhaPavlichenkova-TomTom committed:
format fix

ZuleykhaPavlichenkova-TomTom committed:
Merge branch 'main' into unittester-summary

ZuleykhaPavlichenkova-TomTom committed:
remove unused includes

hmeriann commented:
I've updated the PR 
- added `std::stringstream &GetSummary()` to accumulate the failure logs (if we'll need write them to a file, we could just write them directly from it)
- there is no logging into a file anymore
- in `run_tests_one_by_one.py` it sets `SUMMARIZE_FAILURES` env variable to `'0'` when running the unittest in a subprocess to avoid printing Failures Summary multiple times

ZuleykhaPavlichenkova-TomTom committed:
add function PrintResultString() to work around the result.Print() statement

hmeriann commented:
Windows CI failure is unrelated to the PR

Tmonster reviewed (commented):
Still a couple off comments, we can look together on Friday

hmeriann commented on test/sqlite/sqllogic_test_logger.cpp:
```diff
@@ -1,10 +1,12 @@
 
 namespace duckdb {
 
+static int failures_summary_counter = 1;
```
added a comment and made the counter start with 0, because it's strange start it at 1, indeed)))

ZuleykhaPavlichenkova-TomTom committed:
remove unused relative include
put duplicate strings into a variable where it can be done (except for the strings with termcolor)
format-fix

hmeriann commented on test/sqlite/sqllogic_test_logger.cpp:
```diff
@@ -1,10 +1,12 @@
 #include "duckdb/parser/parser.hpp"
 #include "termcolor.hpp"
-#include "sqllogic_test_runner.hpp"
+#include "../include/test_helpers.hpp"
```
appeared that it is even unused

ZuleykhaPavlichenkova-TomTom committed:
move GetSummary into sqllogic_test_logger and make it static
fix all the appearances of the function to compile

ZuleykhaPavlichenkova-TomTom committed:
ff

hmeriann commented on test/helpers/test_helpers.cpp:
```diff
@@ -27,6 +27,11 @@ static bool single_threaded = false;
 static case_insensitive_set_t required_requires;
 static bool delete_test_path = true;
 
+std::stringstream &GetSummary() {
```
oh, thanks! I've tried this without making it static and it didn't work, but now looks like working

hmeriann commented on test/sqlite/sqllogic_test_logger.cpp:
```diff
@@ -238,6 +274,9 @@ void SQLLogicTestLogger::SplitMismatch(idx_t row_number, idx_t expected_column_c
 	std::cerr << "Expected " << termcolor::bold << expected_column_count << termcolor::reset << " columns, but got "
 	          << termcolor::bold << split_count << termcolor::reset << " columns" << std::endl;
 	std::cerr << "Does the result contain tab values? In that case, place every value on a single row." << std::endl;
+	GetSummary() << "Expected " + to_string(expected_column_count) + " columns, but got " + to_string(split_count) +
```
I've updated the logger where it looked possible to use a `log_str` variable to build a string and output it to `std::cerr` or to `GetSummary()` afterwards. 
It would be great to look together at the places where the `termcolor` is used.

ZuleykhaPavlichenkova-TomTom committed:
Merge branch 'main' into unittester-summary

Tmonster commented on test/sql/copy/csv/test_date.test:
```diff
@@ -11,7 +11,7 @@ CREATE TABLE date_test(d date);
 COPY date_test FROM 'data/csv/test/date.csv';
 ----
-1
+1w
```
? don't think this needs to be in here right?

hmeriann replied:
right!

Tmonster commented on test/sqlite/sqllogic_test_logger.cpp:
```diff
@@ -293,4 +334,9 @@ void SQLLogicTestLogger::LoadDatabaseFail(const string &dbpath, const string &me
 }
 
+std::stringstream &SQLLogicTestLogger::GetSummary() {
+	static std::stringstream summary;
```
Could this be defined in `test/unittest.cpp`? Then it can be called just like `SummarizeFailures()` and all the global summary state is in unitest.cpp just like the `summarize_failures` bool or `test_force_storage` etc.

hmeriann replied:
thank you! It worked

Tmonster reviewed (commented):
Looking much nicer, just one comment about where the summary stream lives

ZuleykhaPavlichenkova-TomTom committed:
revert the test

ZuleykhaPavlichenkova-TomTom committed:
move GetSummary() back to unittest.cpp

ZuleykhaPavlichenkova-TomTom committed:
remove duplicating operations by redirecting all the outputs into a std::stringstream, which accumulates failure messages to output them in std::cerr and later in the summary

ZuleykhaPavlichenkova-TomTom committed:
add GetSummaryCounter() to get track of failed test case order number with start couning from the unittest scope
fix doubled output

ZuleykhaPavlichenkova-TomTom committed:
revert test case

ZuleykhaPavlichenkova-TomTom committed:
Merge branch 'main' into unittester-summary

ZuleykhaPavlichenkova-TomTom committed:
move(log_message) in Log() to probably solve error: use of deleted function issue on GCC 4.8

ZuleykhaPavlichenkova-TomTom committed:
try to summarise when run_test_one_by_one.py a list of tests with an offset

ZuleykhaPavlichenkova-TomTom committed:
cleaning up

ZuleykhaPavlichenkova-TomTom committed:
stringstream object can't be moved in return statement

ZuleykhaPavlichenkova-TomTom committed:
use ostrrringstream to redirect termcolor outputs

hmeriann commented:
`GCC 4.8` CI wasn't happy about a `std::stringstream` as a return type, so I switched back to the `string`.
Then to avoid duplicating lines, the termcolored texts were temporarily buffered in `std::ostringstream oss` and appended to a final string.
This looks now like working and not "eating" a part of the test outputs like it was earlier.

My only concern for now that it collects summary logs into a `std::stringstream` buffer and for handling formatted outputs `std::ostringstream` buffer is used. Does it make sense to make both buffers of the same type or leave it like it is now?

hmeriann commented:
> stringstream should not be a return type

ugh, I've just didn't push my local changes - there are strings are the return types

ZuleykhaPavlichenkova-TomTom committed:
pass std::ostringstream &oss buffer to the logger functions

ZuleykhaPavlichenkova-TomTom committed:
format fix

ZuleykhaPavlichenkova-TomTom committed:
on no_exit it runs tests one by one and they all come with the header created by  PrintSummaryHeader() and each time GetSummaryCounter() in PrintSummaryHeader() starts with 1 and the headers get repeated in the summary. I've tried to work around it in ry script because didn't want to pass additional flags or read env variables in sqllogic_test_logger, but maybe it'd be better on mo_exit to pass env={"NO_EXIT":"1"} and read it in logger

ZuleykhaPavlichenkova-TomTom committed:
avoid duplicating headers in FAILURES SUMMARY for cases with --no-exit and run_tests_batched()

ZuleykhaPavlichenkova-TomTom committed:
format fix

ZuleykhaPavlichenkova-TomTom committed:
allow to print stderr if list_of_tests, add one more case to avoid duplicationg headers

Mytherin commented on test/sqlite/sqllogic_test_logger.cpp:
```diff
@@ -13,45 +13,55 @@ SQLLogicTestLogger::SQLLogicTestLogger(ExecuteContext &context, const Command &c
 		for (idx_t r = 0; r < values.size(); r++) {
 			if (c != 0) {
-				fprintf(stderr, "\t");
+				// fprintf(stderr, "\t");
```
Can we remove the commented out code?

hmeriann replied:
Yes, sure!

Mytherin commented on test/sqlite/sqllogic_test_logger.cpp:
```diff
@@ -13,45 +13,55 @@ SQLLogicTestLogger::SQLLogicTestLogger(ExecuteContext &context, const Command &c
+}
+
+void SQLLogicTestLogger::PrintExpectedResult(const vector<string> &values, idx_t columns, bool row_wise,
+                                             std::ostringstream &oss) {
```
This can just be `std::ostream` - it is a virtual class

hmeriann replied:
I used the `std::ostringstream` because couldn't stream `std::ostream` to the `std::cerr`, because `std::cerr` takes strings, but `std::ostream` doesn't look like have the `str()` method. Can we leave it as `std::ostringstream` or there is a way to make it work with `std::ostream`?

Mytherin commented on test/sqlite/sqllogic_test_logger.cpp:
```diff
@@ -101,196 +111,223 @@ void SQLLogicTestLogger::PrintSQLFormatted() {
-	PrintLineSep();
-	Log("Error message: " + message + "\n");
-	PrintLineSep();
+void SQLLogicTestLogger::LoadDatabaseFail(const string &dbpath, const string &message, std::ostringstream &oss) {
```
Instead of passing this to every method - perhaps the `SQLLogicTestLogger` can just have a `std::ostream &output` that it gets in the constructor, to avoid having to pass this in with all the methods?

hmeriann replied:
I would look cleaner indeed!

I've tried to initialise the `SQLLogicTestLogger` with the `std::stringstream &oss` (which I made a private property of the `SQLLogicTestLogger`) so we could use it without having to pass it in every function. I had to make `SQLLogicTestLogger` methods non-static as well.

Right now I'm having problems with compilation caused by the `SQLLogicTestLogger::LoadDatabaseFail()` called in the `SQLLogicTestRunner` scope. Since the `SQLLogicTestLogger::LoadDatabaseFail()` is not static anymore, it requires an instance to be called from. And I'm failing to create that `logger` instance - cannot figure out what should be passed into the logger's constructor for the `cons Command &command`

hmeriann replied:
Well, I figured out how to make the compiler happy. Should work now

Mytherin reviewed (commented):
Thanks! Looks good - left some more minor comments. There are also some CI failures remaining

ZuleykhaPavlichenkova-TomTom committed:
get access to the tests_per_invocation argument

ZuleykhaPavlichenkova-TomTom committed:
pass std::stringstream &oss in its constuctor, make SQLLogicTestLogger all methods non-static, clean up

ZuleykhaPavlichenkova-TomTom committed:
update pure virtual function Command::ExecuteInternal() - make it just virtual to make compiler happy

ZuleykhaPavlichenkova-TomTom committed:
ff

return empty string if get_test_mane_from(string) didn't find matches

fix test name to be a string, not an array

initialise with guard lock

protecting each ostingstream with the mutex didn't help locally, but lest's try in CI

for the test run

make logger functions static again

ff

ff

only tsan

clean up

proper capture lists for the lambdas, add mutex to the accessing the summary for cerr output

restored from commits used to return strings from logger functions

revert logger to return strings

Tmonster reviewed (commented):
Looks good. Just a couple of questions. Also looks like there is still some CI that needs fixing

hmeriann commented on scripts/run_tests_one_by_one.py:
```diff
@@ -120,6 +123,12 @@ def parse_assertions(stdout):
 
+def get_test_name_from(text):
+    match = re.findall(r'\((.*?)\)\!', text)
+    if match:
```
Yeah, you are right, if there is nothing had matched, it fails on print

hmeriann commented on test/unittest.cpp:
```diff
@@ -24,18 +28,36 @@ bool TestMemoryLeaks() {
 	for (int i = 0; i < argc; i++) {
 		if (string(argv[i]) == "--force-storage") {
 			test_force_storage = true;
-		} else if (string(argv[i]) == "--force-reload" || string(argv[i]) == "--force-restart") {
```
hmmm, I'm not sure, I've put it back

Mytherin commented:
@hmeriann could we pick up work on this again? I would still like to get this in

hmeriann commented:
@Mytherin sure, I'm looking into the Thread Sanitizer errors. Looks like they appeared because I used std::ostringstream at the same time with the std::cerr.

Tishj commented:
> @Mytherin sure, I'm looking into the Thread Sanitizer errors. Looks like they appeared because I used std::ostringstream at the same time with the std::cerr.

I think the ThreadSanitizer output is super confusing because when it triggers an error that isn't ignored (see <https://github.com/duckdb/duckdb/blob/main/.sanitizer-thread-suppressions.txt>) it output all the errors it ignored *as well*.

I'm not 100% on that theory, but looking at the seemingly unrelated issues reported I think that's what's happening.
What I think is happening is that `GetSummary` is the real problem, since it's a static variable, if we have multiple threads running statements at the same time (in for example a `concurrentloop`), the stream could get written to at the same time, which I don't think it's build for.

<https://learn.microsoft.com/en-us/cpp/standard-library/thread-safety-in-the-cpp-standard-library?view=msvc-170>
> If an object is being written to by one thread, then all reads and writes to that object on the same or other threads must be protected. For example, given an object A, if thread 1 is writing to A, then thread 2 must be prevented from reading from or writing to A.

I think the `iostream` currently returned by `GetSummary` should be a class, which has a mutex, and we need to lock that mutex whenever we write to the summary.

Tishj commented on test/unittest.cpp:
```diff
@@ -24,12 +28,30 @@ bool TestMemoryLeaks() {
+}
+
+std::ostringstream &GetSummary() {
+	static std::ostringstream summary;
```
I think we want to wrap this with a mutex:
```c++
struct FailureSummary {
public:
    void AddSummary(const string &failure) {
        std::lock_guard<std::mutex> guard(lock);
        summary << failure;
    }
public:
    std::ostringstream summary;
    std::mutex lock;
};

Tishj replied:
Or, looking at how we use the `<<` operator a lot, we could create this instead:
```c++
void SafeAppend(std::function<void(std::ostringstream &)> callback) {
    std::lock_guard<std::mutex> guard(lock);
    callback(summary);
}
```

used like:
```c++
failure_summary.SafeAppend([](std::ostringstream &oss) {
    oss << termcolor::red << termcolor::bold << "Mismatch on row " << current_row + 1 << ", column "
		    << result.ColumnName(current_column) << "(index " << current_column + 1 << ")" << std::endl
		    << termcolor::reset; 
});
```

Tishj commented on test/sqlite/result_helper.cpp:
```diff
@@ -93,6 +94,7 @@ bool TestResultHelper::CheckQueryResult(const Query &query, ExecuteContext &cont
 		comparison_values = LoadResultFromFile(fname, result.names, expected_column_count, csv_error);
 		if (!csv_error.empty()) {
 			logger.PrintErrorHeader(csv_error);
+			std::cerr << oss.str();
```
This also needs the lock (discussed below)

hmeriann commented:
> `GetSummary` should be a class, which has a mutex, and we need to lock that mutex whenever we write to the summary.

Thank you! This makes a lot of sense (the `SQLLogicTestLogger` has its own mutex, but the stuff I've added doesn't protect the interaction with the stream) and thanks for suggesting the implementation - I've tried it and made it to compile, protected each situation when logger writes to the `std::ostringstream &oss`, but still couldn't make TSAN happy

Tishj commented:
> > `GetSummary` should be a class, which has a mutex, and we need to lock that mutex whenever we write to the summary.
> 
> Thank you! This makes a lot of sense (the `SQLLogicTestLogger` has its own mutex, but the stuff I've added doesn't protect the interaction with the stream) and thanks for suggesting the implementation - I've tried it and made it to compile, protected each situation when logger writes to the `std::ostringstream &oss`, but still couldn't make TSAN happy

I'd be happy to review that code, can you commit it to the branch?

hmeriann commented:
> can you commit it to the branch?

Sure, it would be great if you take another look! I'm looking now for the places where I've overused mutexes

Tishj commented on test/sqlite/sqllogic_test_runner.cpp:
```diff
@@ -104,7 +106,9 @@ void SQLLogicTestRunner::LoadDatabase(string dbpath, bool load_extensions) {
 	} catch (std::exception &ex) {
 		ErrorData err(ex);
-		SQLLogicTestLogger::LoadDatabaseFail(dbpath, err.Message());
+		ExecuteContext context;
```
I'm not sure of the impact of these changes.
We needed this before because we attached the `oss` to the `SQLLogicTestLogger` but we don't do that anymore, so we *could* revert this I think?

hmeriann replied:
yes, you are right

hmeriann replied:
These are some missed to clean up artefacts(((

Tishj commented on test/sqlite/sqllogic_command.hpp:
```diff
@@ -66,7 +66,7 @@ class Command {
 	                                                         string file_name, idx_t query_line) const;
 
-	virtual void ExecuteInternal(ExecuteContext &context) const = 0;
+	virtual void ExecuteInternal(ExecuteContext &context) const;
```
Not related to any issues I think, but is this change necessary?
I believe `Command::ExecuteInternal` is implemented as entirely empty anyways

hmeriann replied:
yes, here it should stay pure virtual since we reverted the `SQLLogicTestLogger` to have static functions again (and the empty implementation should be reverted as well)

Tishj commented on test/include/test_helpers.hpp:
```diff
@@ -58,6 +64,27 @@ void WriteBinary(string path, const uint8_t *data, uint64_t length);
+};
+
+inline FailureSummary &GetFailureSummary() {
+	static FailureSummary instance;
```
This is just a hunch, but perhaps it's a problem that GetFailureSummary is implemented in the header, since it's using a static

I'm not that well versed on statics and globals, but I know it's a problem if a global is defined in a header, because then it gets copied to every cpp file that includes it, defeating the purpose of the global

hmeriann replied:
Moving an implementation from the header file is a good idea. And I think you are right about the static FailureSummary - don't know what else can still don't let getting rid of the data race...

I think maybe we could break down FailuresSummary accumulation like:
1. Change return types of logger functions to std::string
2. Stream them to std::cerr
3. Stream them into a static string(?) protecting with the mutex

ZuleykhaPavlichenkova-TomTom committed:
move Safe append and Get methods to the logger, create a local statix mutex and a vector of strings instead of the string

Tmonster commented on scripts/run_tests_one_by_one.py:
```diff
@@ -139,14 +166,19 @@ def launch_test(test, list_of_tests=False):
             test_cmd = ['valgrind'] + test_cmd
-        res = subprocess.run(test_cmd, stdout=unittest_stdout, stderr=unittest_stderr, timeout=timeout)
+        # should unset SUMMARIZE_FAILURES to avoid producing exceeding failure logs
+        if list_of_tests or no_exit or tests_per_invocation:
```
try the following lines instead
```py
        if args.valgrind:
            test_cmd = ['valgrind'] + test_cmd
        # should unset SUMMARIZE_FAILURES to avoid producing exceeding failure logs
        env = os.environ.copy()
        if list_of_tests or no_exit or tests_per_invocation:
            env['SUMMARIZE_FAILURES'] = '0'
            env['NO_DUPLICATING_HEADERS'] = '1'
        else:
            env['SUMMARIZE_FAILURES'] = '0'
        res = subprocess.run(test_cmd, stdout=unittest_stdout, stderr=unittest_stderr, timeout=timeout, env=env)
```

The default behavior for subprocess.run is to pass the current environment to the subprocesses unless env is passed. You are passing a new environment on line 181, which means the current environment is no longer passed, so you loose all of the thread suppression environment variables. If you copy the whole environment, it should work

ZuleykhaPavlichenkova-TomTom committed:
add a helper function to python runner to clean up the message captured from stderr

ZuleykhaPavlichenkova-TomTom committed:
Merge branch 'main' of https://github.com/duckdb/duckdb into unittester-summary

ZuleykhaPavlichenkova-TomTom committed:
add back what was accidentially removed from the WrongResultHash() signature

ZuleykhaPavlichenkova-TomTom committed:
when there is no "===" in the error output, we'd like the get_clean_error_message_from(text) to return whole text

ZuleykhaPavlichenkova-TomTom committed:
add missing header

ZuleykhaPavlichenkova-TomTom committed:
Merge remote-tracking branch 'origin/main' into unittester-summary

ZuleykhaPavlichenkova-TomTom committed:
compare inputs to strings

Mytherin commented on test/helpers/test_helpers.cpp:
```diff
@@ -404,4 +404,35 @@ bool compare_result(string csv, ColumnDataCollection &collection, vector<Logical
+// 	summary += line;
+// }
+
+// std::string FailureSummary::ToString() const {
```
Can we either clear out the commented code or use the commented out code?

Mytherin commented on test/sqlite/sqllogic_test_logger.cpp:
```diff
@@ -13,45 +18,85 @@ SQLLogicTestLogger::SQLLogicTestLogger(ExecuteContext &context, const Command &c
+	failures_summary.push_back(log_message);
+}
+
+void SQLLogicTestLogger::LogBoth(const string &log_message) {
```
Can we just rename this method to `LogFailure`? The fact that it writes to two destinations is more of a detail of the method.

Mytherin commented on test/sqlite/sqllogic_test_logger.cpp:
```diff
@@ -13,45 +18,85 @@ SQLLogicTestLogger::SQLLogicTestLogger(ExecuteContext &context, const Command &c
-	std::cerr << str;
+void SQLLogicTestLogger::SafeAppend(const string &log_message) {
+	lock_guard<mutex> lock(summary_mutex);
+	failures_summary.push_back(log_message);
```
`SafeAppend` is not clear in what it does either - maybe call this method `AppendFailure`?

hmeriann replied:
`AppendFailure` makes more sense indeed, thank you!

Mytherin reviewed (commented):
Can we pick this up again? After digging through some more logs this is still very much a needed change. I've left some more comments:

ZuleykhaPavlichenkova-TomTom committed:
cleaning up

ZuleykhaPavlichenkova-TomTom committed:
Format UnexpectedFailure message like all other failures

ZuleykhaPavlichenkova-TomTom committed:
ff

ZuleykhaPavlichenkova-TomTom committed:
SafeAppend() => AppendFailure()

ZuleykhaPavlichenkova-TomTom committed:
remove commented code

ZuleykhaPavlichenkova-TomTom committed:
Merge branch 'main' into unittester-summary

ZuleykhaPavlichenkova-TomTom committed:
no duplicating headers

ZuleykhaPavlichenkova-TomTom committed:
add SUMMARIZE_FAILURES as step's env

Mytherin commented on test/include/test_helpers.hpp:
```diff
@@ -25,19 +25,26 @@
 #include "duckdb/common/string_util.hpp"
 #include "duckdb/common/enum_util.hpp"
 #include "duckdb/common/types.hpp"
+#include "../sqlite/sqllogic_test_logger.hpp"
```
Can we remove this include?

Mytherin commented on test/sqlite/sqllogic_test_runner.cpp:
```diff
@@ -92,6 +92,8 @@ void SQLLogicTestRunner::EndLoop() {
 void SQLLogicTestRunner::LoadDatabase(string dbpath, bool load_extensions) {
 	loaded_databases.push_back(dbpath);
 
+	// auto &oss = GetSummary();
```
nit: commented out code

Mytherin commented on test/sqlite/sqllogic_test_runner.cpp:
```diff
@@ -104,6 +106,7 @@ void SQLLogicTestRunner::LoadDatabase(string dbpath, bool load_extensions) {
 		ExtensionHelper::LoadExtension(*db, "core_functions");
 	} catch (std::exception &ex) {
 		ErrorData err(ex);
+		ExecuteContext context;
```
This seems unused?

Mytherin commented on test/sqlite/sqllogic_test_logger.cpp:
```diff
@@ -1,10 +1,15 @@
 namespace duckdb {
 
+static mutex summary_mutex;
+static vector<string> failures_summary;
```
Instead of having this be global statics can we move this to a separate class (`FailureSummary` or similar)?

Mytherin commented on test/unittest.cpp:
```diff
@@ -11,6 +12,10 @@ namespace duckdb {
 static bool test_memory_leaks = false;
+static bool summarize_failures = false;
+// this counter is for the order number of the failed test case in Failures Summary
+static size_t failures_summary_counter = 0;
```
Same here - can we move all of this to a single class that holds all of the required state? The class can either be a singleton or just have static methods.

Mytherin reviewed (commented):
Thanks for the fixes! Looks good. Some more comments below. I also noticed that we are not reporting time out failures in `run_tests_one_by_one.py`, see e.g. the "Release Assertions with Clang"  run. Can we report those as well?

> [2573/4670]: test/sql/join/asof/test_asof_join_missing.test_slowStill running...
Still running...
Still running...
 (TIMED OUT)

ZuleykhaPavlichenkova-TomTom committed:
Clean up and organise the Failures Summary static variables into a class

ZuleykhaPavlichenkova-TomTom committed:
strategy fail-fast applies to matrix, but we don't have a matrix in Code Coverage job

ZuleykhaPavlichenkova-TomTom committed:
FailureSeummary as a class
